File: tools/dagtoml-validate-go/runtime_document.go
package main

import (
	"fmt"
	"path/filepath"
	"reflect"
	"strings"
)

// rdArray preserves every element, including malformed scalar entries.
func rdArray(value any) ([]any, bool) {
	if value == nil {
		return nil, false
	}
	v := reflect.ValueOf(value)
	if v.Kind() != reflect.Slice {
		return nil, false
	}
	out := make([]any, v.Len())
	for i := range out {
		out[i] = v.Index(i).Interface()
	}
	return out, true
}

func rdVocab(root, attribute string) ([]string, bool, bool) {
	paths, err := filepath.Glob(filepath.Join(root, "profiles", "*", "ontology.toml"))
	if err != nil || !containsString(paths, filepath.Join(root, "profiles/agent-assurance/ontology.toml")) {
		return nil, false, false
	}
	paths = append(paths, filepath.Join(root, "core/ontology.toml"))
	seen := map[string]bool{}
	var found []string
	var foundExtensible, foundOK bool
	for _, path := range paths {
		doc, err := loadDoc(path)
		if err != nil {
			return nil, false, false
		}
		entries, ok := rdArray(doc["attribute_vocabularies"])
		if !ok || len(entries) == 0 {
			return nil, false, false
		}
		for _, raw := range entries {
			entry, ok := raw.(map[string]any)
			if !ok {
				return nil, false, false
			}
			name, nameOK := entry["attribute"].(string)
			extensible, extOK := entry["extensible"].(bool)
			values, valuesOK := rdArray(entry["values"])
			if !nameOK || name == "" || seen[name] || !extOK || !valuesOK {
				return nil, false, false
			}
			seen[name] = true
			tokens := []string{}
			for _, value := range values {
				token, ok := value.(string)
				if !ok || containsString(tokens, token) {
					return nil, false, false
				}
				tokens = append(tokens, token)
			}
			if name == attribute {
				found, foundExtensible, foundOK = tokens, extensible, true
			}
		}
	}
	return found, foundExtensible, foundOK
}

func validateRuntimeFields(doc rawDoc, kind, root string) []string {
	var errors []string
	meta, _ := tableOf(doc, "meta")
	if meta["template_kind"] != kind {
		errors = append(errors, "RDMETA: meta.template_kind must match the selected kind")
	}
	profile, _ := meta["framework_profile"].(string)
	if profile != "agent-assurance" && profile != "AGDF" {
		errors = append(errors, "RDMETA: meta.framework_profile must identify agent-assurance")
	}
	bindings := []struct {
		kind, section, field, attribute string
		required                        bool
	}{
		{"adapter-contract", "adapter", "runtime_kind", "runtime_kind", true},
		{"adapter-contract", "adapter", "runtime_network_policy", "runtime_network_policy", true},
		{"adapter-contract", "adapter", "runtime_clock_policy", "runtime_clock_policy", true},
		{"adapter-contract", "adapter", "id_derivation", "adapter_id_derivation", false},
		{"adapter-registry-binding", "binding", "adapter_ref_syntax", "adapter_ref_syntax", true},
		{"gate-decision", "decision", "verdict", "gate_decision_verdict", true},
	}
	for _, b := range bindings {
		if b.kind != kind {
			continue
		}
		table, ok := tableOf(doc, b.section)
		if !ok {
			errors = append(errors, fmt.Sprintf("RDVOCAB %s: %s must be a table", b.attribute, b.section))
			continue
		}
		value, present := table[b.field]
		if !b.required && !present {
			continue
		}
		values, extensible, loaded := rdVocab(root, b.attribute)
		if !loaded || extensible || len(values) == 0 {
			errors = append(errors, "RDLOAD: missing/invalid closed vocabulary "+b.attribute)
			continue
		}
		token, stringOK := value.(string)
		if !stringOK || !containsString(values, token) {
			errors = append(errors, fmt.Sprintf("RDVOCAB %s: %s.%s must be a declared string", b.attribute, b.section, b.field))
		}
	}
	return errors
}

func rdNonempty(table map[string]any, field, location string, errors *[]string) {
	value, ok := table[field].(string)
	if !ok || strings.TrimSpace(value) == "" {
		*errors = append(*errors, fmt.Sprintf("RDSTRUCT: %s.%s must be a nonempty string", location, field))
	}
}

func rdEntries(table map[string]any, field, location string, required bool, errors *[]string) []map[string]any {
	value, present := table[field]
	if !present && !required {
		return nil
	}
	entries, ok := rdArray(value)
	if !ok || (required && len(entries) == 0) {
		*errors = append(*errors, fmt.Sprintf("RDSTRUCT: %s.%s must be an array of tables%s", location, field, map[bool]string{true: " with at least one entry", false: ""}[required]))
		return nil
	}
	var out []map[string]any
	for _, entry := range entries {
		row, ok := entry.(map[string]any)
		if !ok {
			*errors = append(*errors, fmt.Sprintf("RDSTRUCT: every %s.%s entry must be a table", location, field))
			continue
		}
		out = append(out, row)
	}
	return out
}

func validateRuntimeAdapter(doc rawDoc, root string, binding bool) []string {
	kind, section := "adapter-contract", "adapter"
	if binding {
		kind, section = "adapter-registry-binding", "binding"
	}
	errors := validateRuntimeFields(doc, kind, root)
	table, ok := tableOf(doc, section)
	if !ok {
		return errors
	}
	if binding {
		rdNonempty(table, "adapter_ref", section, &errors)
		rdNonempty(table, "registry_url", section, &errors)
		scheme, schemeOK := table["registry_url_scheme"].(string)
		url, urlOK := table["registry_url"].(string)
		values, _, loaded := rdVocab(root, "registry_scheme")
		if !loaded {
			errors = append(errors, "RDLOAD: invalid registry_scheme declaration")
		} else if !schemeOK || !containsString(values, scheme) {
			errors = append(errors, "INV01: binding.registry_url_scheme must be declared in registry_scheme")
		}
		if !schemeOK || !urlOK || !strings.HasPrefix(url, scheme+"://") {
			errors = append(errors, "INV02: binding.registry_url must start with the declared scheme followed by ://")
		}
		for _, entry := range rdEntries(table, "trust_anchor_refs", section, false, &errors) {
			rdNonempty(entry, "anchor_id", "binding.trust_anchor_refs[]", &errors)
		}
		for _, entry := range rdEntries(table, "policy_constraint_refs", section, false, &errors) {
			value, ok := entry["constraint_id"].(string)
			if !ok || !gdAssertionRe.MatchString(value) {
				errors = append(errors, "INV03: binding.policy_constraint_refs[].constraint_id must match the whole ASCII assertion-ID grammar")
			}
		}
	} else {
		rdNonempty(table, "input_source", section, &errors)
		values, extensible, loaded := rdVocab(root, "input_hash_method")
		if !loaded {
			return append(errors, "RDLOAD: invalid input_hash_method declaration")
		}
		if raw, present := table["input_hash_method"]; present {
			value, ok := raw.(string)
			if !ok || (!extensible && !containsString(values, value)) {
				errors = append(errors, "RDVOCAB input_hash_method: adapter.input_hash_method must satisfy its declared string vocabulary")
			}
		}
		runtimeKind, _ := table["runtime_kind"].(string)
		field := map[string]string{"wasi-component": "component_digest", "oci-action": "oci_image_digest", "os-sandbox": "profile_args_digest"}[runtimeKind]
		artifact, ok := table["runtime_artifact"].(map[string]any)
		if !ok {
			errors = append(errors, "INV02: adapter.runtime_artifact must be a table")
		} else if field != "" {
			value, ok := artifact[field].(string)
			if !ok || !gdHex64Re.MatchString(value) {
				errors = append(errors, "INV02: adapter.runtime_artifact."+field+" must be exactly 64 lowercase ASCII hex characters")
			}
		}
		for _, entry := range rdEntries(table, "emits", section, true, &errors) {
			value, ok := entry["ijb_primitive"].(string)
			if !ok || !containsString([]string{"thing", "scope", "path", "observed", "constraint", "time"}, value) {
				errors = append(errors, "INV03: adapter.emits[].ijb_primitive must be one of the six IJB primitives")
			}
		}
		for _, entry := range rdEntries(table, "conformance_fixtures", section, true, &errors) {
			for _, field := range []string{"raw_input_ref", "expected_bundle_ref"} {
				rdNonempty(entry, field, "adapter.conformance_fixtures[]", &errors)
			}
		}
		for _, entry := range rdEntries(table, "declared_invariants", section, false, &errors) {
			for _, field := range []string{"id", "description", "enforced_by"} {
				rdNonempty(entry, field, "adapter.declared_invariants[]", &errors)
			}
		}
		if value, present := table["runtime_env_allowlist"]; present {
			entries, ok := rdArray(value)
			valid := ok
			for _, entry := range entries {
				if _, ok := entry.(string); !ok {
					valid = false
				}
			}
			if !valid {
				errors = append(errors, "RDSTRUCT: adapter.runtime_env_allowlist must be an array of strings")
			}
		}
	}
	return errors
}

File: tools/dagtoml-validate-rs/src/runtime_document.rs
//! SPEC structural checks for document-owned runtime declarations. No execution.

use super::*;

pub fn vocab(repo_root: &Path, attribute: &str) -> Option<(Vec<String>, bool)> {
    let mut paths = vec![repo_root.join("core/ontology.toml")];
    let profile_root = repo_root.join("profiles");
    for entry in std::fs::read_dir(profile_root).ok()? {
        let path = entry.ok()?.path().join("ontology.toml");
        if path.is_file() {
            paths.push(path);
        }
    }
    if !paths.contains(&repo_root.join("profiles/agent-assurance/ontology.toml")) {
        return None;
    }
    paths.sort();
    let mut seen = std::collections::BTreeSet::new();
    let mut result = None;
    for path in paths {
        let doc = load(&path).ok()?;
        let entries = doc.get("attribute_vocabularies")?.as_array()?;
        if entries.is_empty() {
            return None;
        }
        for entry in entries {
            let name = entry.get("attribute")?.as_str()?;
            if name.is_empty() || !seen.insert(name.to_owned()) {
                return None;
            }
            let extensible = entry.get("extensible")?.as_bool()?;
            let values = entry.get("values")?.as_array()?;
            let mut tokens = Vec::new();
            for value in values {
                let token = value.as_str()?.to_owned();
                if tokens.contains(&token) {
                    return None;
                }
                tokens.push(token);
            }
            if name == attribute {
                result = Some((tokens, extensible));
            }
        }
    }
    result
}

pub fn hex64(value: &str) -> bool {
    value.len() == 64
        && value
            .bytes()
            .all(|b| b.is_ascii_digit() || (b'a'..=b'f').contains(&b))
}

fn assertion_id(value: &str) -> bool {
    let Some(tail) = value.strip_prefix("A-") else {
        return false;
    };
    let mut bytes = tail.bytes();
    bytes.next().is_some_and(|b| b.is_ascii_alphanumeric())
        && bytes.all(|b| b.is_ascii_alphanumeric() || b == b'_' || b == b'-')
}

pub fn fields(doc: &Value, kind: &str, repo_root: &Path) -> Vec<String> {
    let mut errors = Vec::new();
    let meta = doc.get("meta");
    if meta
        .and_then(|m| m.get("template_kind"))
        .and_then(Value::as_str)
        != Some(kind)
    {
        errors.push(format!("RDMETA: meta.template_kind must equal {kind:?}"));
    }
    let profile = meta
        .and_then(|m| m.get("framework_profile"))
        .and_then(Value::as_str);
    if !matches!(profile, Some("agent-assurance" | "AGDF")) {
        errors.push("RDMETA: meta.framework_profile must identify agent-assurance".into());
    }
    let bindings = [
        (
            "adapter-contract",
            "adapter",
            "runtime_kind",
            "runtime_kind",
            true,
        ),
        (
            "adapter-contract",
            "adapter",
            "runtime_network_policy",
            "runtime_network_policy",
            true,
        ),
        (
            "adapter-contract",
            "adapter",
            "runtime_clock_policy",
            "runtime_clock_policy",
            true,
        ),
        (
            "adapter-contract",
            "adapter",
            "id_derivation",
            "adapter_id_derivation",
            false,
        ),
        (
            "adapter-registry-binding",
            "binding",
            "adapter_ref_syntax",
            "adapter_ref_syntax",
            true,
        ),
        (
            "gate-decision",
            "decision",
            "verdict",
            "gate_decision_verdict",
            true,
        ),
    ];
    for (owner, section, field, attribute, required) in bindings {
        if kind != owner {
            continue;
        }
        let Some(table) = doc.get(section).and_then(Value::as_table) else {
            errors.push(format!("RDVOCAB {attribute}: {section} must be a table"));
            continue;
        };
        if !required && !table.contains_key(field) {
            continue;
        }
        let Some((values, false)) = vocab(repo_root, attribute) else {
            errors.push(format!(
                "RDLOAD: missing/invalid closed vocabulary {attribute}"
            ));
            continue;
        };
        if values.is_empty()
            || !table
                .get(field)
                .and_then(Value::as_str)
                .is_some_and(|v| values.iter().any(|x| x == v))
        {
            errors.push(format!(
                "RDVOCAB {attribute}: {section}.{field} must be a declared string"
            ));
        }
    }
    errors
}

fn nonempty(
    table: &toml::map::Map<String, Value>,
    field: &str,
    location: &str,
    errors: &mut Vec<String>,
) {
    if !table
        .get(field)
        .and_then(Value::as_str)
        .is_some_and(|v| !v.trim().is_empty())
    {
        errors.push(format!(
            "RDSTRUCT: {location}.{field} must be a nonempty string"
        ));
    }
}

fn entries<'a>(
    table: &'a toml::map::Map<String, Value>,
    field: &str,
    location: &str,
    required: bool,
    errors: &mut Vec<String>,
) -> Vec<&'a toml::map::Map<String, Value>> {
    if !required && !table.contains_key(field) {
        return Vec::new();
    }
    let Some(values) = table.get(field).and_then(Value::as_array) else {
        errors.push(format!(
            "RDSTRUCT: {location}.{field} must be an array of tables"
        ));
        return Vec::new();
    };
    if required && values.is_empty() {
        errors.push(format!("RDSTRUCT: {location}.{field} must be nonempty"));
    }
    let mut result = Vec::new();
    for value in values {
        match value.as_table() {
            Some(entry) => result.push(entry),
            None => errors.push(format!(
                "RDSTRUCT: every {location}.{field} entry must be a table"
            )),
        }
    }
    result
}

pub fn validate(doc: &Value, repo_root: &Path, binding: bool) -> Vec<String> {
    let kind = if binding {
        "adapter-registry-binding"
    } else {
        "adapter-contract"
    };
    let section = if binding { "binding" } else { "adapter" };
    let mut errors = fields(doc, kind, repo_root);
    let Some(table) = doc.get(section).and_then(Value::as_table) else {
        return errors;
    };
    if binding {
        nonempty(table, "adapter_ref", section, &mut errors);
        nonempty(table, "registry_url", section, &mut errors);
        let scheme = table.get("registry_url_scheme").and_then(Value::as_str);
        match vocab(repo_root, "registry_scheme") {
            Some((values, _)) if scheme.is_some_and(|s| values.iter().any(|v| v == s)) => (),
            Some(_) => errors.push(
                "INV01: binding.registry_url_scheme must be declared in registry_scheme".into(),
            ),
            None => errors.push("RDLOAD: invalid registry_scheme declaration".into()),
        }
        let matches = scheme
            .zip(table.get("registry_url").and_then(Value::as_str))
            .is_some_and(|(s, url)| url.starts_with(&format!("{s}://")));
        if !matches {
            errors.push(
                "INV02: binding.registry_url must start with the declared scheme followed by ://"
                    .into(),
            );
        }
        for entry in entries(table, "trust_anchor_refs", section, false, &mut errors) {
            nonempty(
                entry,
                "anchor_id",
                "binding.trust_anchor_refs[]",
                &mut errors,
            );
        }
        for entry in entries(table, "policy_constraint_refs", section, false, &mut errors) {
            if !entry
                .get("constraint_id")
                .and_then(Value::as_str)
                .is_some_and(assertion_id)
            {
                errors.push("INV03: binding.policy_constraint_refs[].constraint_id must match the whole ASCII assertion-ID grammar".into());
            }
        }
    } else {
        nonempty(table, "input_source", section, &mut errors);
        let Some((values, extensible)) = vocab(repo_root, "input_hash_method") else {
            errors.push("RDLOAD: invalid input_hash_method declaration".into());
            return errors;
        };
        if let Some(value) = table.get("input_hash_method") {
            if !value
                .as_str()
                .is_some_and(|s| extensible || values.iter().any(|v| v == s))
            {
                errors.push("RDVOCAB input_hash_method: adapter.input_hash_method must satisfy its declared string vocabulary".into());
            }
        }
        let digest = match table.get("runtime_kind").and_then(Value::as_str) {
            Some("wasi-component") => Some("component_digest"),
            Some("oci-action") => Some("oci_image_digest"),
            Some("os-sandbox") => Some("profile_args_digest"),
            _ => None,
        };
        match table.get("runtime_artifact").and_then(Value::as_table) {
            None => errors.push("INV02: adapter.runtime_artifact must be a table".into()),
            Some(artifact) => {
                if let Some(field) = digest {
                    if !artifact
                        .get(field)
                        .and_then(Value::as_str)
                        .is_some_and(hex64)
                    {
                        errors.push(format!("INV02: adapter.runtime_artifact.{field} must be exactly 64 lowercase ASCII hex characters"));
                    }
                }
            }
        }
        for entry in entries(table, "emits", section, true, &mut errors) {
            if !matches!(
                entry.get("ijb_primitive").and_then(Value::as_str),
                Some("thing" | "scope" | "path" | "observed" | "constraint" | "time")
            ) {
                errors.push(
                    "INV03: adapter.emits[].ijb_primitive must be one of the six IJB primitives"
                        .into(),
                );
            }
        }
        for entry in entries(table, "conformance_fixtures", section, true, &mut errors) {
            for field in ["raw_input_ref", "expected_bundle_ref"] {
                nonempty(entry, field, "adapter.conformance_fixtures[]", &mut errors);
            }
        }
        for entry in entries(table, "declared_invariants", section, false, &mut errors) {
            for field in ["id", "description", "enforced_by"] {
                nonempty(entry, field, "adapter.declared_invariants[]", &mut errors);
            }
        }
        if let Some(env) = table.get("runtime_env_allowlist") {
            if !env
                .as_array()
                .is_some_and(|values| values.iter().all(Value::is_str))
            {
                errors.push(
                    "RDSTRUCT: adapter.runtime_env_allowlist must be an array of strings".into(),
                );
            }
        }
    }
    errors
}
